src/platform/datapath_raw_xdp.c

```diff
@@ -788,6 +803,93 @@ CxPlatDpRawInterfaceUpdateRules(
     }
 }
 
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+CxPlatDpRawInterfaceAddRule(
+    _In_ XDP_INTERFACE* Interface,
+    _In_ const XDP_RULE* NewRule
+    )
+{
+#pragma warning(push)
+#pragma warning(disable:6386) // Buffer overrun while writing to 'NewRules' - FALSE POSITIVE
+
+    CxPlatLockAcquire(&Interface->RuleLock);
+    // TODO - Don't always allocate a new array?
+
+    if (Interface->RuleCount + 1 == 0) {
+        QuicTraceEvent(
+            LibraryError,
+            "[ lib] ERROR, %s.",
+            "No more room for rules");
+        CxPlatLockRelease(&Interface->RuleLock);
+        return;
+    }
+
+    const size_t OldSize = sizeof(XDP_RULE) * (size_t)Interface->RuleCount;
+    const size_t NewSize = sizeof(XDP_RULE) * ((size_t)Interface->RuleCount + 1);
+
+    XDP_RULE* NewRules = CxPlatAlloc(NewSize, RULE_TAG);
```

Comment:
Why does QUIC need to maintain a list of rules? It looks like you're only plumbing one rule per queue, so why can't you just call XdpCreateProgram and stash the returned program handle in the XDP_QUEUE, and then close that handle when the queue is deleted (or when a new rule is created)?

Reply:
It's appending the new rule to the array.

Reply:
I'm asking why QUIC needs to maintain this array.

Reply:
Because XDP doesn't support the notion of "add rule" or "remove rule". We have to plumb them all down together.

Reply:
>>It looks like you're only plumbing one rule per queue
One rule per socket from my understanding. There could be multiple rules on a queue.

Reply:
Correct. Each socket has 1 rule, but they all apply to the same queue(s).

Reply:
We need to have a brownbag on the object model here.